composer/callbacks/speed_monitor.py

```diff
@@ -108,11 +140,34 @@ def batch_end(self, state: State, logger: Logger):
         self.batch_wct_buffer.append(batch_wct)
         self.batch_num_samples_buffer.append(batch_num_samples)
 
-        # Log the throughput
         if len(self.batch_num_samples_buffer) == self.window_size:
+            # Log the throughput
             throughput = sum(self.batch_num_samples_buffer) / sum(self.batch_wct_buffer)
             logger.log_metrics({'throughput/samples_per_sec': throughput})
 
+            # Estimate remaining time
+            batch_wct_avg = sum(self.batch_wct_buffer) / len(self.batch_wct_buffer)
+            elapsed_dur = self.get_elapsed_duration(state)
+            if elapsed_dur is None:
+                warnings.warn('`max_duration` is not set. Cannot estimate remaining time.')
+            elif elapsed_dur > 0 and elapsed_dur != self.last_elapsed_duration:
+                # Only update the estimate if the elapsed duration has changed
+                self.last_elapsed_duration = elapsed_dur
+                remaining_time = batch_wct_avg * int(state.timestamp.batch) / elapsed_dur * (1 - elapsed_dur)
+                # Add remaining time from each evaluator
+                for dataloader_label, eval_wcts in self.eval_wct_per_label.items():
+                    # Discard first eval_wct if possible as it often slower due to dataset downloading
+                    eval_wct_avg = None
```

Comment:
eh... I want to push back on this since shortcutting a len call seems a bit unnecessary and it adds another level of indirection when tracing code. open to discussion tho

Reply:
If I was asking you to call this `len_eval_wcts = len(eval_wcts)`, then I think that would be unnecessary indirection, but in this case the variable is to make it more clear what this term in the equation means i.e. if you were to explain out exactly why the formula is as it is, would you refer to this as the length of the `eval_wcts` list or as the number of times eval has happened? A comment could work too if you don't want to do a variable name

Reply:
changed my mind, will go with ur original suggestion